x-pack/filebeat/input/cometd/input_test.go

```diff
@@ -20,182 +21,274 @@ import (
+		event1 := <-eventsCh
+		assertEventMatches(t, expected, event1)
 
-	cfg := common.MustNewConfigFrom(config)
+		// run input
+		input2.Run()
+		defer input2.Stop()
 
-	input, err := NewInput(cfg, connector, inputContext)
-	require.NoError(t, err)
-	require.NotNil(t, input)
+		event2 := <-eventsCh
+		assertEventMatches(t, expected, event2)
+
+		// close server
+		server.Close()
+	})
 
-	input.Run()
+	t.Run("Check gracefully shutdown input", func(t *testing.T) {
+		eventsCh := make(chan beat.Event)
 
-	go func() {
-		time.Sleep(100 * time.Millisecond) // let input.Stop() be executed.
-		for range eventsCh {
+		outlet := &mockedOutleter{
+			onEventHandler: func(event beat.Event) bool {
+				eventsCh <- event
+				return true
+			},
+		}
+		connector := &mockedConnector{
+			outlet: outlet,
+		}
+		var inputContext finput.Context
+
+		var msg bay.MaybeMsg
+		msg.Msg.Data.Event.ReplayID = 1234
+		msg.Msg.Data.Payload = []byte(`{"CountryIso": "IN"}`)
+		msg.Msg.Channel = firstChannel
+
+		config := map[string]interface{}{
+			"channel_name":              firstChannel,
+			"auth.oauth2.client.id":     "client.id",
+			"auth.oauth2.client.secret": "client.secret",
+			"auth.oauth2.user":          "user",
+			"auth.oauth2.password":      "password",
 		}
-	}()
 
-	input.Wait()
-}
+		r := http.HandlerFunc(oauth2Handler)
+		server := httptest.NewServer(r)
+		serverURL = server.URL
+		config["auth.oauth2.token_url"] = serverURL + "/token"
 
-func TestStop(t *testing.T) {
-	conf := defaultConfig()
-	logger := logp.NewLogger("test")
-	authParams := bay.AuthenticationParameters{}
-	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
-	workerCtx, workerCancel := context.WithCancel(inputCtx)
-	defer cancelInputCtx()
-
-	input := &cometdInput{
-		config:       conf,
-		log:          logger,
-		inputCtx:     inputCtx,
-		workerCtx:    workerCtx,
-		workerCancel: workerCancel,
-		ackedCount:   atomic.NewUint32(0),
-		authParams:   authParams,
-	}
-	input.out = make(chan bay.TriggerEvent)
+		cfg := common.MustNewConfigFrom(config)
 
-	input.Stop()
-}
+		input, err := NewInput(cfg, connector, inputContext)
+		require.NoError(t, err)
+		require.NotNil(t, input)
 
-func TestWait(t *testing.T) {
-	conf := defaultConfig()
-	logger := logp.NewLogger("test")
-	authParams := bay.AuthenticationParameters{}
-	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
-	workerCtx, workerCancel := context.WithCancel(inputCtx)
-	defer cancelInputCtx()
-
-	input := &cometdInput{
-		config:       conf,
-		log:          logger,
-		inputCtx:     inputCtx,
-		workerCtx:    workerCtx,
-		workerCancel: workerCancel,
-		ackedCount:   atomic.NewUint32(0),
-		authParams:   authParams,
-	}
-	input.out = make(chan bay.TriggerEvent)
+		// run input
+		input.Run()
+
+		go func() {
+			time.Sleep(100 * time.Millisecond) // let input.Stop() be executed.
+			input.Wait()
+		}()
 
-	input.Wait()
+		for range []beat.Event{<-eventsCh} {
+		}
```

Comment:
```suggestion
        select {
		case _, ok := <-eventsCh:
		    require.False(ok, "Expected channel to be closed")
		case <- time.After(time.Second):
		     require.Fail("Channel was not closed after Wait()")
		}
```
What is this for loop trying to do? Are you trying to wait for the channel to be closed? In that case something like the above would be clearer.

If this doesn't work it would be better to have an explicit timeout than to wait forever at least.

Reply:
Hey @cmacknz,
it's not working. tests hang in between. Let me see if I can use timeout here somehow.